src/accordion/AccordionPanel.ts

```diff
@@ -1,55 +1,140 @@
+  unstable_IdHTMLProps,
 } from "reakit";
+import * as React from "react";
+import { useForkRef } from "reakit-utils";
+import { createHook, createComponent } from "reakit-system";
 
 import { ACCORDION_PANEL_KEYS } from "./__keys";
 import { AccordionStateReturn } from "./AccordionState";
-import { useAccordionItemContext } from "./AccordionItem";
 
-export type AccordionPanelOptions = unstable_IdOptions &
-  Pick<AccordionStateReturn, "registerPanel" | "items" | "activeItems">;
+export type AccordionPanelOptions = DisclosureContentOptions &
+  unstable_IdOptions &
+  Pick<
+    AccordionStateReturn,
+    | "selectedId"
+    | "selectedIds"
+    | "registerPanel"
+    | "unregisterPanel"
+    | "panels"
+    | "items"
+    | "allowMultiple"
+  > & {
+    /**
+     * Accordion's id
+     */
+    accordionId?: string;
+  };
 
-export type AccordionPanelHTMLProps = unstable_IdHTMLProps;
+export type AccordionPanelHTMLProps = DisclosureContentHTMLProps &
+  unstable_IdHTMLProps;
 
 export type AccordionPanelProps = AccordionPanelOptions &
   AccordionPanelHTMLProps;
 
+function getAccordionsWithoutPanel(
+  accordions: AccordionPanelOptions["items"],
+  panels: AccordionPanelOptions["panels"],
+) {
+  const panelsAccordionIds = panels.map(panel => panel.groupId).filter(Boolean);
+
+  return accordions.filter(
+    item => panelsAccordionIds.indexOf(item.id || undefined) === -1,
+  );
+}
+
+function getPanelIndex(
+  panels: AccordionPanelOptions["panels"],
+  panel: typeof panels[number],
+) {
+  const panelsWithoutAccordionId = panels.filter(p => !p.groupId);
+  return panelsWithoutAccordionId.indexOf(panel);
+}
+
+/**
+ * When <AccordionPanel> is used without accordionId:
+ *
+ *  - First render: getAccordionId will return undefined because options.panels
+ * doesn't contain the current panel yet (registerPanel wasn't called yet).
+ * Thus registerPanel will be called without groupId (accordionId).
+ *
+ *  - Second render: options.panels already contains the current panel (because
+ * registerPanel was called in the previous render). This means that we'll be
+ * able to get the related accordionId with the accordion panel index. Basically,
+ * we filter out all the accordions and panels that have already matched. In this
+ * phase, registerPanel will be called again with the proper groupId (accordionId).
+ *
+ *  - In the third render, panel.groupId will be already defined, so we just
+ * return it. registerPanel is not called.
+ */
+function getAccordionId(options: AccordionPanelOptions) {
+  const panel = options.panels?.find(p => p.id === options.id);
+  const accordionId = options.accordionId || panel?.groupId;
+  if (accordionId || !panel || !options.panels || !options.items) {
+    return accordionId;
+  }
+
+  const panelIndex = getPanelIndex(options.panels, panel);
+  const accordionsWithoutPanel = getAccordionsWithoutPanel(
+    options.items,
+    options.panels,
+  );
+  return accordionsWithoutPanel[panelIndex]?.id || undefined;
+}
+
```

Comment:
Move these functions to the end since isPanelVisible is also on the end.

Reply:
Okay 👍